virtcontainers/device/manager/utils.go

```diff
@@ -35,3 +40,92 @@ func isVFIO(hostPath string) bool {
 func isBlock(devInfo config.DeviceInfo) bool {
 	return devInfo.DevType == "b"
 }
+
+func IsVFIOLargeBarSpaceDevice(hostPath string) (bool, error) {
+	if !isVFIO(hostPath) {
+		return false, nil
+	}
+	iommuDevicesPath := filepath.Join(config.SysIOMMUPath, filepath.Base(hostPath), "devices")
+	deviceFiles, err := ioutil.ReadDir(iommuDevicesPath)
+	if err != nil {
+		return false, err
+	}
+	// Pass all devices in iommu group
+	for _, deviceFile := range deviceFiles {
+		//For example, 0000:04:00.0
+		tokens := strings.Split(deviceFile.Name(), ":")
+		vfioDeviceType := config.VFIODeviceErrorType
+		if len(tokens) == 3 {
+			vfioDeviceType = config.VFIODeviceNormalType
+		} else {
+			//For example, 83b8f4f2-509f-382f-3c1e-e6bfe0fa1001
+			tokens = strings.Split(deviceFile.Name(), "-")
+			if len(tokens) == 5 {
+				vfioDeviceType = config.VFIODeviceMediatedType
+			}
+		}
+		deviceFileName := filepath.Join(iommuDevicesPath, deviceFile.Name())
+		var isLarge bool
+		switch vfioDeviceType {
+		case config.VFIODeviceNormalType:
+			if isLarge, err = isLargeBarSpace(filepath.Join(deviceFileName, "resource")); err != nil {
+				return false, err
+			}
+			logrus.WithFields(logrus.Fields{
+				"device-file":        deviceFile.Name(),
+				"device-type":        vfioDeviceType,
+				"is-large-bar-space": isLarge,
+			}).Info("Detect large bar space device")
+			return isLarge, nil
+		case config.VFIODeviceMediatedType:
+			logrus.WithFields(logrus.Fields{
+				"device-file": deviceFile.Name(),
+				"device-type": vfioDeviceType,
+			}).Warn("Detect large bar space device is not yet supported for VFIODeviceMediatedType")
```

Comment:
@Jimmy-Xu Any reason why this is not supported for mediated devices?

Reply:
I have no test environment currently :(